cmd/skopeo/proxy.go

```diff
@@ -0,0 +1,574 @@
+		return ret, fmt.Errorf("failed to load image: %w", err)
+	}
+
+	h.imageSerial++
+	openimg := openImage{
+		id:  h.imageSerial,
+		src: imgsrc,
+		img: img,
+	}
+	h.images[openimg.id] = openimg
+	ret.value = openimg.id
+
+	return ret, nil
+}
+
+func (h *proxyHandler) CloseImage(args []interface{}) (replyBuf, error) {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+	var ret replyBuf
+
+	if h.sysctx == nil {
+		return ret, fmt.Errorf("Must invoke Initialize")
+	}
+	if len(args) != 1 {
+		return ret, fmt.Errorf("invalid request, expecting one argument")
+	}
+	imgref, err := h.parseImageFromID(args[0])
+	if err != nil {
+		return ret, err
+	}
+	imgref.src.Close()
+	delete(h.images, imgref.id)
+
+	return ret, nil
+}
+
+func parseImageID(v interface{}) (uint32, error) {
+	imgidf, ok := v.(float64)
+	if !ok {
+		return 0, fmt.Errorf("Expecting integer imageid, not %T", v)
+	}
+	return uint32(imgidf), nil
+}
+
+func (h *proxyHandler) parseImageFromID(v interface{}) (*openImage, error) {
+	imgid, err := parseImageID(v)
+	if err != nil {
+		return nil, err
+	}
+	imgref, ok := h.images[imgid]
+	if !ok {
+		return nil, fmt.Errorf("No image %v", imgid)
+	}
+	return &imgref, nil
+}
+
+func (h *proxyHandler) allocPipe() (*os.File, *activePipe, error) {
+	piper, pipew, err := os.Pipe()
+	if err != nil {
+		return nil, nil, err
+	}
+	f := activePipe{
+		w: pipew,
+	}
+	h.activePipes[uint32(pipew.Fd())] = f
+	f.wg.Add(1)
+	return piper, &f, nil
+}
+
+// GetManifest returns a copy of the manifest, converted to OCI format, along with the original digest.
+func (h *proxyHandler) GetManifest(args []interface{}) (replyBuf, error) {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	var ret replyBuf
+
+	if h.sysctx == nil {
+		return ret, fmt.Errorf("Must invoke Initialize")
+	}
+	if len(args) != 1 {
+		return ret, fmt.Errorf("invalid request, expecting one argument")
+	}
+	imgref, err := h.parseImageFromID(args[0])
+	if err != nil {
+		return ret, err
+	}
+
+	ctx := context.TODO()
+	rawManifest, _, err := imgref.img.Manifest(ctx)
+	if err != nil {
+		return ret, err
+	}
+	// We always return the original digest, as that's what clients need to do pull-by-digest
+	// and in general identify the image.
+	digest, err := manifest.Digest(rawManifest)
+	if err != nil {
+		return ret, err
+	}
+	// But, we convert to OCI format on the wire.  The idea here is that by reusing the containers/image
+	// stack, clients to this proxy can pretend the world is OCI only, and not need to care about e.g.
+	// docker schema.
+	tempdest, err := os.MkdirTemp("", "skopeo-proxy-tmp-oci")
+	if err != nil {
+		return ret, err
+	}
+	defer os.RemoveAll(tempdest)
+	fakeociref, err := ocilayout.ParseReference(tempdest)
+	if err != nil {
+		return ret, err
+	}
+	fakeocidest, err := fakeociref.NewImageDestination(context.Background(), h.sysctx)
+	if err != nil {
+		return ret, err
+	}
+
+	manifestUpdates := types.ManifestUpdateOptions{InformationOnly: types.ManifestUpdateInformation{Destination: fakeocidest}}
```

Comment:
> and, alternatively, if the client in question doesn’t care about the config

This is the case for now - we don't need the config because we're not actually running these as containers when we go to import into the host ostree.  Or to say it another way, the config has no data we care about.

(Except, I may want to fetch this for the history annotations at least since that's really useful to render...and there is a really natural desire to support re-exporting an image, but I'm trying to avoid scope-creeping into that for now)

> , I find it a bit surprising that it cares so much about getting the full manifest data).

Well, you're right that my code today just needs the layers.  But perhaps we e.g. render annotations in the future.  

Since OCI is both simple (easy to parse) and standardized, my thought was that clients can parse that instead of a separarate custom "Layer" struct like happens internally in c/image.

I admit the separation of duty here between client and proxy is somewhat ad-hoc.  But it's been working?